Implemented first stab at thread safety

kristoiv opened the pull request:
Stab at resolving issue #13 . Looking for feedback.

kristoiv commented:
Oh, and tests run fine.

maxekman commented on fsm.go:
```diff
@@ -125,6 +129,7 @@ type Callbacks map[string]Callback
 // currently performed.
 func NewFSM(initial string, events []EventDesc, callbacks map[string]Callback) *FSM {
 	var f FSM
+	f.mutex = sync.RWMutex{}
```
This is not needed, it is ready to go as it is.

kristoiv replied:
Fixed

maxekman commented on fsm.go:
```diff
@@ -42,6 +43,9 @@ type transitionerStruct struct {
 // It has to be created with NewFSM to function properly.
 type FSM struct {
+	// mutex is used for synchronization to achieve thread safety
+	mutex sync.RWMutex
```
Would be cleaner to have it as a anonymous field.

kristoiv replied:
If it was anonymous, wouldn't the Lock/RLock/Unlock/RUnlock methods be public? Thus external access would be dangerous.

maxekman replied:
Yes, you are correct! Keep the current approach.

maxekman commented on fsm.go:
```diff
@@ -199,23 +204,31 @@ func NewFSM(initial string, events []EventDesc, callbacks map[string]Callback) *
 
 // Current returns the current state of the FSM.
 func (f *FSM) Current() string {
+	f.mutex.RLock()
```
With a anonymous field this would simply be `f.RLock()`.

maxekman commented on fsm.go:
```diff
@@ -286,6 +302,13 @@ func (f *FSM) Event(event string, args ...interface{}) error {
+}
+
+// transitionInternal wraps transitioner.transition.
+func (f *FSM) transitionInternal() error {
```
I think `doTransition()` would be a better name.

kristoiv replied:
Fixed

kristoiv commented:
What do we think now?

kristoiv commented:
Ah, I didn't remember your first comment. Creating a test-case may be difficult since it is a race condition.

One problem I see with the new implementation is access from callbacks will block if they use the public API of their own *FSM object.

maxekman commented:
A test case using goroutine for either the read or write together with the `-race` flag should trigger it. See https://golang.org/doc/articles/race_detector.html

Yes, I was also thinking a bit by the possible case of deadlocking within callbacks. Not sure how to prevent that right now. Do you have ideas?

kristoiv commented:
I have written a testcase that results in a deadlock, trying to figure out a clean way of doing it. But there will be a lot more lock/unlock calls during Event(...)'s.

maxekman commented:
Wonderful! Let me know when I should take a look.

kristoiv commented:
There! I went back to before I added thread safety and added a test proving the lack of thread safety, then I merged it to my current HEAD and confirmed the problem to be solved.

I also fixed the deadlock problem as well as I could.

kristoiv commented:
Cool, you do that :)

maxekman commented on fsm.go:
```diff
@@ -306,13 +326,17 @@ func (t transitionerStruct) transition(f *FSM) error {
 // general version.
 func (f *FSM) beforeEventCallbacks(e *Event) error {
 	if fn, ok := f.callbacks[cKey{e.Event, callbackBeforeEvent}]; ok {
+		f.mutex.Unlock()
```
It would be less repetitive and error prone in these cases to wrap the callbacks in `NewFSM` at line 193 with a function that does the unlocking and locking:
```
f.callbacks[cKey{target, callbackType}] = func(e *Event) {
    // This always happens in a locked condition, but needs to be unlocked
    // to not risk a deadlock. The reversed order is not a mistake!
    f.mutex.Unlock()
    defer f.mutex.Lock()
    c(e)
}
```
If you do this, could you also change `var f FSM` in the beginning to pointer form:
```
f := &FSM{
	transitionerObj: &transitionerStruct{},
	current: initial,
	transitions: make(map[eKey]string),
	callbacks: make(map[cKey]Callback),
}
```
And return it with out dereferencing.

Do you think this would be good?

kristoiv replied:
Yes definitely. I fixed both. Good call!

kristoiv commented:
Ops: I simplified my solution a little more

kristoiv commented:
I added your comment :)

In our case defering or not doesn't make any difference, but I changed it as you suggest. Arguably if the code is ever changed, whether or not it should be run in a lock or not depends on its function.

maxekman commented:
Definitely, I can be too picky sometimes!

Do you think we should merge this now?

kristoiv commented:
Haha, it's a good looking syntax in general to keep locks and unlocks together, so why not 👍 .
I don't see why not, looks to be a good improvement.

maxekman commented:
Thanks for all the work with this PR!

kristoiv commented:
No problem. Thanks for all your feedback!
